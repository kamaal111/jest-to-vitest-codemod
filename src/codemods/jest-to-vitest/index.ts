import fs from 'node:fs/promises';
import type { Dirent } from 'node:fs';
import path from 'node:path';

import { Lang, parseAsync, type SgNode, type SgRoot } from '@ast-grep/napi';
import type { Kinds, TypesMap } from '@ast-grep/napi/types/staticTypes.js';
import type { Codemod, Modifications, RunCodemodOkResult } from '@kamaalio/codemod-kit';
import { type types, objects } from '@kamaalio/kamaal';

import hasAnyJestGlobalAPI from './utils/has-any-jest-global-api.js';
import replaceJestApiWithVi from './rules/replace-jest-api-with-vi.js';
import jestHooksToVitest from './rules/jest-hooks-to-vitest.js';
import jestMockTypeToVitest from './rules/jest-mock-type-to-vitest.js';
import addVitestImports from './rules/add-vitest-imports.js';
import removeJestImport from './rules/remove-jest-import.js';

export const JEST_TO_VITEST_LANGUAGE = Lang.TypeScript;

const ESLINT_VERSION = '^3.2.4';

function jestToVitestFilter(root: SgNode<TypesMap, Kinds<TypesMap>>): boolean {
  return hasAnyJestGlobalAPI(root);
}

export async function jestToVitestModifications(modifications: Modifications): Promise<Modifications> {
  const root = modifications.ast.root();
  if (!jestToVitestFilter(root)) return modifications;

  return replaceJestApiWithVi(modifications)
    .then(jestHooksToVitest)
    .then(jestMockTypeToVitest)
    .then(addVitestImports)
    .then(removeJestImport);
}

async function jestToVitest(content: SgRoot<TypesMap> | string, filename?: types.Optional<string>): Promise<string> {
  const ast = typeof content === 'string' ? await parseAsync(JEST_TO_VITEST_LANGUAGE, content) : content;

  return jestToVitestModifications(makeJestToVitestInitialModification(ast, filename)).then(modifications => {
    return modifications.ast.root().text();
  });
}

export function makeJestToVitestInitialModification(
  ast: SgRoot<TypesMap>,
  filename: types.Optional<string> = null,
): Modifications {
  return {
    lang: JEST_TO_VITEST_LANGUAGE,
    report: { changesApplied: 0 },
    ast,
    filename,
    history: [ast],
  };
}

async function jestToVitestPostTransform(
  {
    root,
    results,
  }: {
    root: string;
    results: Array<RunCodemodOkResult>;
  },
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _codemod: Codemod,
): Promise<void> {
  if (results.length === 0) return;

  let content: Array<Dirent<string>>;
  try {
    content = await fs.readdir(root, { withFileTypes: true });
  } catch {
    return;
  }

  const existingVitestConfig = content.find(item => item.isFile() && item.name.startsWith('vitest.config.'));
  if (existingVitestConfig != null) return;

  const vitestConfigContent = `import { defineConfig } from 'vitest/config';

export default defineConfig({
  // Configure Vitest (https://vitest.dev/config/)
  test: {},
});
  `.trim();

  await fs.writeFile(path.join(root, 'vitest.config.ts'), vitestConfigContent);

  let packageJsonContent: string;
  try {
    packageJsonContent = await fs.readFile(path.join(root, 'package.json'), { encoding: 'utf-8' });
  } catch {
    // Just a project without a package.json
    return;
  }

  const packageJson = JSON.parse(packageJsonContent);
  const existingVitestDependency = packageJson.devDependencies?.vitest;
  if (existingVitestDependency != null) return;

  const devDependencies = objects.omitBy(
    Object.fromEntries(
      Object.entries({
        ...(packageJson.devDependencies ?? {}),
        vitest: ESLINT_VERSION,
      }).sort(([a], [b]) => a.localeCompare(b)),
    ),
    item => item == null,
  );
  const updatedPackageJson = { ...packageJson, devDependencies };
  await fs.writeFile(path.join(root, 'package.json'), JSON.stringify(updatedPackageJson, null, 2) + '\n');
}

export const JEST_TO_VITEST_CODEMOD: Codemod = {
  name: 'jest-to-vitest-transformer',
  languages: [JEST_TO_VITEST_LANGUAGE],
  transformer: jestToVitest,
  postTransform: jestToVitestPostTransform,
};

export default jestToVitest;
