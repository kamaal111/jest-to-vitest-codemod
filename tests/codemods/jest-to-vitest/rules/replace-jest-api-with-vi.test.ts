import { describe, expect, it } from 'vitest';

import { invalidRuleSignal, validRuleSignal } from '../../../test-utils/detection-theory';
import { JEST_TO_VITEST_LANGUAGE, makeJestToVitestInitialModification } from '../../../../src/codemods/jest-to-vitest';
import replaceJestApiWithVi from '../../../../src/codemods/jest-to-vitest/rules/replace-jest-api-with-vi';

describe('jest.requireActual -> vi.importActual', () => {
  it('replaces jest.requireActual with vi.importActual', async () => {
    const source = `jest.requireActual('lodash/cloneDeep')`;

    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`requireActual`);
    expect(updatedSource).toContain(`vi.importActual('lodash/cloneDeep')`);
  });

  it('replaces jest.requireActual with vi.importActual in mock', async () => {
    const source = `
    jest.mock('something', () => ({
      ...jest.requireActual('something'),
      mocked: jest.fn()
    }))
    `;

    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`requireActual`);
    expect(updatedSource).toContain(`await vi.importActual('something')`);
    expect(updatedSource).toContain(`vi.mock('something', async () => ({`);
  });
});

describe('jest.mock -> vi.mock', () => {
  it('replaces jest mock with vi', async () => {
    const source = `jest.mock('./some-path', () => 'hello')`;

    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).toContain(`vi.mock('./some-path', () => ({`);
    expect(updatedSource).toContain(`default: 'hello'`);
  });

  it('replaces jest mock with vi without module override', async () => {
    const source = `
    jest.mock('./some-path')
    `;

    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).toContain(`vi.mock('./some-path')`);
  });

  it('does not replace anything', async () => {
    const source = `
    vi.mock('./some-path', () => ({
        default: 'hello',
    }))
    `;

    await validRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
  });
});

describe('jest.spyOn -> vi.spyOn', () => {
  it('replaces jest spyOn with vi', async () => {
    const source = `jest.spyOn(modules, 'path', 'key')`;

    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`jest.spyOn`);
    expect(updatedSource).toContain(`vi.spyOn(modules, 'path', 'key')`);
  });

  it('replaces jest spyOn with vi with a single param', async () => {
    const source = `jest.spyOn(modules)`;

    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`jest.spyOn`);
    expect(updatedSource).toContain(`vi.spyOn(modules)`);
  });
});

describe('jest.restoreAllMocks -> vi.restoreAllMocks', () => {
  it('replaces jest restoreAllMocks with vi', async () => {
    const source = `jest.restoreAllMocks()`;
    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`jest.restoreAllMocks`);
    expect(updatedSource).toContain(`vi.restoreAllMocks()`);
  });
});

describe('jest.resetAllMocks -> vi.resetAllMocks', () => {
  it('replaces jest resetAllMocks with vi', async () => {
    const source = `jest.resetAllMocks()`;
    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`jest.resetAllMocks`);
    expect(updatedSource).toContain(`vi.resetAllMocks()`);
  });
});

describe('jest.clearAllMocks -> vi.clearAllMocks', () => {
  it('replaces jest clearAllMocks with vi', async () => {
    const source = `jest.clearAllMocks()`;
    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`jest.clearAllMocks`);
    expect(updatedSource).toContain(`vi.clearAllMocks()`);
  });
});

describe('jest.useFakeTimers -> vi.useFakeTimers', () => {
  it('replaces jest useFakeTimers with vi', async () => {
    const source = `jest.useFakeTimers()`;
    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`jest.useFakeTimers`);
    expect(updatedSource).toContain(`vi.useFakeTimers()`);
  });
});

describe('jest.useRealTimers -> vi.useRealTimers', () => {
  it('replaces jest useRealTimers with vi', async () => {
    const source = `jest.useRealTimers()`;
    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`jest.useRealTimers`);
    expect(updatedSource).toContain(`vi.useRealTimers()`);
  });
});

describe('jest.setTimeout -> vi.setTimeout', () => {
  it('replaces jest setTimeout with vi', async () => {
    const source = `jest.setTimeout(50_000)`;
    const modifications = await invalidRuleSignal(source, JEST_TO_VITEST_LANGUAGE, ast => {
      return replaceJestApiWithVi(makeJestToVitestInitialModification(ast));
    });
    const updatedSource = modifications.ast.root().text();

    expect(updatedSource).not.toContain(`jest.setTimeout`);
    expect(updatedSource).toContain(`vi.setTimeout({ testTimeout: 50_000 })`);
  });
});
