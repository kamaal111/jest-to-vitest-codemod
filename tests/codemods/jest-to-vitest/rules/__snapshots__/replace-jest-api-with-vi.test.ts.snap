// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jest.clearAllMocks -> vi.clearAllMocks > replaces jest clearAllMocks with vi 1`] = `"vi.clearAllMocks()"`;

exports[`jest.mock -> vi.mock > replaces jest mock with vi 1`] = `"vi.mock('./some-path', () => ({ default: 'hello' }))"`;

exports[`jest.mock -> vi.mock > replaces jest mock with vi without module override 1`] = `"vi.mock('./some-path')"`;

exports[`jest.requireActual -> vi.importActual > replaces jest.requireActual with vi.importActual 1`] = `"(await vi.importActual('lodash/cloneDeep'))"`;

exports[`jest.requireActual -> vi.importActual > replaces jest.requireActual with vi.importActual in mock 1`] = `
"vi.mock('something', async () => ({ default: ({
      ...(await vi.importActual('something')),
      mocked: jest.fn()
    }) }))"
`;

exports[`jest.resetAllMocks -> vi.resetAllMocks > replaces jest resetAllMocks with vi 1`] = `"vi.resetAllMocks()"`;

exports[`jest.restoreAllMocks -> vi.restoreAllMocks > replaces jest restoreAllMocks with vi 1`] = `"vi.restoreAllMocks()"`;

exports[`jest.setTimeout -> vi.setTimeout > replaces jest setTimeout with vi 1`] = `"vi.setTimeout({ testTimeout: 50_000 })"`;

exports[`jest.spyOn -> vi.spyOn > replaces jest spyOn with vi 1`] = `"vi.spyOn(modules, 'path', 'key')"`;

exports[`jest.spyOn -> vi.spyOn > replaces jest spyOn with vi with a single param 1`] = `"vi.spyOn(modules)"`;

exports[`jest.useFakeTimers -> vi.useFakeTimers > replaces jest useFakeTimers with vi 1`] = `"vi.useFakeTimers()"`;

exports[`jest.useRealTimers -> vi.useRealTimers > replaces jest useRealTimers with vi 1`] = `"vi.useRealTimers()"`;
